ytt/001-schemas/README.md

```diff
@@ -0,0 +1,1647 @@
+  - if this argument is omited, the key specified in the document must match that in the schema.
+- `expects=Int|String|List|Function` — (optional) expected number of items to be `allowed`
+  (and that are not matched by other schema).
+   - `Int` — must match this number, exactly
+   - `String` (e.g. `"1+"`) — must match _at least_ the number
+   - `Function(found):Bool` — predicate of whether the expected number of key were found
+      - `found` (`Int`) — number of actual matches
+   - `List[Int|String|Function]` — must match one of the given criteria
+   - Default: `1` (`Int`) (i.e. match exactly one (1) item)
+- `missing_ok=Bool` — (optional) short-hand for `expects="0+"`
+
+Notes:
+- this annotation is only applicable to map items. If it is specified on any other kind of node,
+  it is an error.
+- if a map item satisfies more than one `@schema/key allowed`, an error results. This is to
+  avoid unintended interactions of multiple wildcard-like rules.
+- additional string-based values for `allowed` may be discovered and included as short-hand for
+  commonly-used predicates.
+- the re-use of `expects` from Overlays is deliberate: in both cases, the parameter captures
+  how many of a given item are expected.
+
+_Example 1: Required with any key_
+
+To require a map item with unspecified key, containing an array of strings:
+
+```yaml
+connection_options:
+  #@schema/key allowed="any"
+  _: 
+  - ""
+```
+
+__
+
+_Example 2: Optional with any key_
+
+To specify an optional map item with unspecified key, containing an array of strings:
+
+```yaml
+connection_options:
+  #@schema/key allowed="any" expects=[0,1]
+  _: 
+  - ""
+```
+
+__
+
+_Example 3: Any number of keys_
+
+To specify any number of map items with unspecified keys, each containing an array of strings:
+
+```yaml
+connection_options:
+  #@schema/key allowed="any" expects="1+"
+  _: 
+  - ""
+```
+
+__
+
+_Example 4: Specific possible keys_
+
+To specify a map item with a key of a specific set, containing an array of strings:
+
+```yaml
+connection_options:
+  #@schema/key allowed=lambda key,_: key in ["aws", "gcp", "azure"]
+  _: 
+  - ""
+```
+
+__
+
+_Example 5: Required suffix
+
+To specify a map item whose key must have a specific suffix, each containing an array of strings:
+
+```yaml
+connection_options:
+  #@schema/key allowed=lambda key, suffix: key.endswith(suffix)
+  _db: 
+  - ""
+```
+
+__
+
+_Example 6: Optional key
+
+To specify an optional map item. If present, must match the specified key and shape.
+
+```yaml
+connection_options:
+  #@schema/key missing_ok=True
+  bucket: 
+    access_key: ""
+    secret_key: ""
+```
+
+## Part 6: Documenting Schema
+
+Configuration Authors may document individual items in a Schema to clarify their meaning, purpose, and impact.
+They may also wish to decorate schema with descriptive information that can generate documentation or a User Interface
+designed to ease configuration.
+
```

Comment:
@cppforlife notes that the way in which Data Values are documented is a core aspect of the UX of `ytt`. The documentation portion of the Schema feature must strive to meet the documentation formatting needs of users, out of the box.

Consider how programming languages include documentation generation as a first-class feature.